engine/src/main/scala/cromwell/jobstore/JobStoreWriterActor.scala

```diff
@@ -1,96 +1,89 @@
 
 import akka.actor.{ActorRef, LoggingFSM, Props}
-import cromwell.jobstore.JobStoreActor._
 import cromwell.core.Dispatcher.EngineDispatcher
+import cromwell.core.actor.BatchingDbWriter
+import cromwell.core.actor.BatchingDbWriter._
+import cromwell.jobstore.JobStore.{Completion, JobCompletion, WorkflowCompletion}
+import cromwell.jobstore.JobStoreActor._
 
+import scala.concurrent.duration._
 import scala.util.{Failure, Success}
+import JobStoreWriterActor._
+
+import scala.language.postfixOps
 
-/**
-  * Singleton actor to coordinate writing job statuses to the database.
-  *
-  * State: Represents an actor either doing nothing, or currently writing to the database
-  * Data: If currently writing, the actor stores pending updates in the data. When one write completes, any further writes are written
-  */
-case class JobStoreWriterActor(jsd: JobStore) extends LoggingFSM[JobStoreWriterState, JobStoreWriterData] {
+
+case class JobStoreWriterActor(jsd: JobStore, dbBatchSize: Int, flushRate: FiniteDuration) extends LoggingFSM[BatchingDbWriterState, BatchingDbWriter.BatchingDbWriterData] {
 
   implicit val ec = context.dispatcher
 
-  startWith(Pending, JobStoreWriterData.empty)
+  startWith(WaitingToWrite, NoData)
 
-  when(Pending) {
-    case Event(command: JobStoreWriterCommand, stateData) =>
-      val newData = writeNextOperationToDatabase(stateData.withNewOperation(sender, command))
-      goto(WritingToDatabase) using newData
+  override def preStart(): Unit = {
+    context.system.scheduler.schedule(0.seconds, flushRate, self, ScheduledFlushToDb)
+    super.preStart()
   }
 
-  when(WritingToDatabase) {
-    case Event(command: JobStoreWriterCommand, stateData) =>
-      stay using stateData.withNewOperation(sender, command)
-    case Event(WriteComplete, stateData) =>
-      val newData = writeNextOperationToDatabase(stateData)
-      goto(if (newData.isEmpty) Pending else WritingToDatabase) using newData
+  when(WaitingToWrite) {
+    case Event(command: JobStoreWriterCommand, curData) =>
+      curData.addData(CommandAndReplyTo(command, sender)) match {
+        case newData: HasData[_] if newData.length >= dbBatchSize => goto(WritingToDb) using newData
+        case newData => stay() using newData
+      }
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic job store flush to DB")
+      goto(WritingToDb) using curData
   }
 
-  whenUnhandled {
-    case Event(someMessage, stateData) =>
-      log.error(s"JobStoreWriter: Unexpected message received in state $stateName: $someMessage")
-      stay()
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay
+    case Event(command: JobStoreWriterCommand, curData) => stay using curData.addData(CommandAndReplyTo(command, sender))
+    case Event(FlushBatchToDb, NoData) =>
+      log.debug("Attempted job store flush to DB but had nothing to write")
+      goto(WaitingToWrite)
+    case Event(FlushBatchToDb, HasData(data)) =>
```

Comment:
ToL: If we're going to block workflow completion behind job store writing, we want to make sure this batch rate is fast enough, otherwise we're going to have annoying wait times at the end of our workflows while we wait for the job store writing batch to happen